### PR TITLE
[IOTDB-5555] Enable modify external RPC EndPoint of DataNode

### PR DESCRIPTION
**Pull request by CRZbulabula**
The cluster DataNode is now enable to modify its clientRPCEndPoint.

The [cluster Node startup protocol](https://apache-iotdb.feishu.cn/docx/BWAtdxAZ8oflr7xcJXncAAyAnMc) is updated for adapting this new feature.

An IT is added for testing this new feature.

You can find the modification result either by SQL show datanodes or in the system.properties file:

image

image

image

### CODE REVIEW

**Review by SpriCoder — APPROVED** [review]
LGTM~

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/node/ClusterNodeStartUtils.java [hunk 1]
```diff
@@ -195,30 +195,39 @@ public static TSStatus confirmNodeRestart(
...
+        }
         break;
       case DataNode:
       default:
-        isTEndPointUpdated =
-            isTEndPointsOfTDataNodeLocationUpdated(
+        updatedTEndPoints =
+            checkUpdatedTEndPointOfDataNode(
                 (TDataNodeLocation) nodeLocation, (TDataNodeLocation) matchedNodeLocation);
+        if (!updatedTEndPoints.isEmpty()
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/node/ClusterNodeStartUtils.java]: Fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/node/ClusterNodeStartUtils.java [hunk 2]
```diff
@@ -336,47 +345,54 @@ public static TDataNodeLocation matchRegisteredDataNode(
   /**
    * Check if some TEndPoints of the specified ConfigNode have updated.
    *
-   * @return True if some TEndPoints of the specified ConfigNode have updated, false otherwise.
+   * @param restartLocation The location of restart ConfigNode
+   * @param recordLocation The record ConfigNode location
+   * @return The set of TEndPoints that have modified. 0: internalEndPoint, 1: consensusEndPoint
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/node/ClusterNodeStartUtils.java]: Removed

File: confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/write/datanode/UpdateDataNodePlan.java [hunk 3]
```diff
@@ -30,42 +31,49 @@
...
   public boolean equals(Object o) {
-    if (this == o) return true;
-    if (o == null || getClass() != o.getClass()) return false;
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    if (!super.equals(o)) {
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/write/datanode/UpdateDataNodePlan.java]: Fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/write/datanode/UpdateDataNodePlan.java [hunk 4]
```diff
@@ -30,42 +31,49 @@
...
+    }
     UpdateDataNodePlan that = (UpdateDataNodePlan) o;
-    return dataNodeLocation.equals(that.dataNodeLocation);
+    return dataNodeConfiguration.equals(that.dataNodeConfiguration);
   }
 
   @Override
   public int hashCode() {
-    return Objects.hash(dataNodeLocation);
+    return Objects.hash(super.hashCode(), dataNodeConfiguration);
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/write/datanode/UpdateDataNodePlan.java]: Fixed~

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Pull request merged**